File: selenium-comparison/spec/tests/pd01_product_info_display_spec.rb
require_relative '../spec_helper'

# Mirrors: tests/product/pd-01-product-info-display.spec.ts
# Guest session — product detail page is publicly accessible.

APEX_ROOT = '/ords/r/sanjay_sikder/ecommerceportal'

RSpec.describe '4 — Product Detail Page' do
  before do
    @driver.navigate.to("#{ENV.fetch('BASE_URL')}#{APEX_ROOT}/product-detail-info?pid_product=1")
    Selenium::WebDriver::Wait.new(timeout: 15).until do
      @driver.execute_script('return document.readyState') == 'complete'
    end
    wait_for_apex_idle(@driver, timeout: 20)
    # Wait for the product content to render — APEX loads it via AJAX after page ready
    Selenium::WebDriver::Wait.new(timeout: 20).until do
      @driver.find_elements(:xpath, "//a[contains(.,'Samsung')]").any?(&:displayed?)
    end
  rescue Selenium::WebDriver::Error::TimeoutError
    nil # proceed — the assertion will produce the meaningful failure
  end

  it 'PD-01 — Product detail page displays full product information' do
    expect(@driver.title).to eq('Product Detail Info')

    # Samsung product name visible as a link (use . to match text in child spans too)
    samsung_links = @driver.find_elements(:xpath, "//a[contains(.,'Samsung')]")
    expect(samsung_links.any?(&:displayed?)).to be(true)

    # Full price 40,000.00 visible somewhere on the page
    price_els = @driver.find_elements(:xpath, "//*[contains(text(),'40,000.00')]")
    expect(price_els.any?(&:displayed?)).to be(true)

    # Product image with alt "Default Logo"
    img = @driver.find_elements(:xpath, "//img[@alt='Default Logo']")
    expect(img.any?(&:displayed?)).to be(true)

    # Review List table with expected column headers
    review_table = @driver.find_elements(:xpath,
      "//table[@aria-label='Review List'] | //table[@summary='Review List']"
    ).first
    expect(review_table).not_to be_nil
    expect(review_table.displayed?).to be(true)

    %w[Client\ User Review\ Message Review\ Date User\ Rating].each do |col|
      # Use . (full descendant text) not text() (direct text nodes only) — APEX wraps th content in spans
      headers = @driver.find_elements(:xpath, "//th[contains(.,'#{col}')]")
      expect(headers.any?(&:displayed?)).to be(true), "Column '#{col}' not visible"
    end

    # >= 3 rows (1 header + >= 2 data rows)
    rows = review_table.find_elements(:tag_name, 'tr')
    expect(rows.length).to be >= 3

    # Statistics region visible with expected cells
    stats_region = @driver.find_elements(:xpath,
      "//*[@role='region'][@aria-label='Statistics'] | //*[@aria-label='Statistics']"
    ).first
    expect(stats_region&.displayed?).to be(true)

    ['Product Rate', 'Total Chatting or Query', 'Total Review'].each do |cell_text|
      cells = @driver.find_elements(:xpath, "//td[contains(.,'#{cell_text}')]")
      expect(cells.any?(&:displayed?)).to be(true), "Cell '#{cell_text}' not visible"
    end

    # Rating Chart region visible
    rating_chart = @driver.find_elements(:xpath,
      "//*[@role='region'][@aria-label='Rating Chart'] | //*[@aria-label='Rating Chart']"
    ).first
    expect(rating_chart&.displayed?).to be(true)
  end
end

File: tests/product/pd-01-product-info-display.spec.ts
// spec: specs/sanzu.plan.md
// seed: fixtures/fixtures.ts

import { test, expect } from '../../fixtures/fixtures';

test.describe('4 — Product Detail Page', () => {
  test('PD-01 — Product detail page displays full product information', async ({ productDetailPage, page }) => {
    // Step 1: Guest. Navigate to /product-detail-info?pid_product=1
    await page.goto(productDetailPage.path + '?pid_product=1');
    // expect: Page title is 'Product Detail Info'
    await expect(page).toHaveTitle('Product Detail Info');

    // Step 2: Assert productName is visible; text matches 'Samsung'
    // expect: Product name visible and equals 'Samsung'
    await expect(page.getByRole('link', { name: 'Samsung' })).toBeVisible();

    // Step 3: Assert price row 'Price-মূল্য' shows '40,000.00'
    // expect: Full price '40,000.00' visible in details table
    await expect(page.getByText('40,000.00')).toBeVisible();

    // Step 4: Assert product image is visible (img element near Add To Cart button)
    // expect: Product image rendered and visible
    await expect(page.getByRole('img', { name: 'Default Logo' })).toBeVisible();

    // Step 5: Assert Review List table visible with expected columns
    // expect: Review table present
    const reviewTable = page.getByRole('table', { name: 'Review List' });
    await expect(reviewTable).toBeVisible();
    await expect(page.getByRole('columnheader', { name: 'Client User' })).toBeVisible();
    await expect(page.getByRole('columnheader', { name: 'Review Message' })).toBeVisible();
    await expect(page.getByRole('columnheader', { name: 'Review Date' })).toBeVisible();
    await expect(page.getByRole('columnheader', { name: 'User Rating' })).toBeVisible();
    // expect: Samsung has >= 2 review rows
    const reviewRows = reviewTable.getByRole('row');
    expect(await reviewRows.count()).toBeGreaterThanOrEqual(3); // header + 2 data rows

    // Step 6: Assert Statistics region visible with expected rows
    // expect: Statistics table visible with expected rows
    await expect(page.getByRole('region', { name: 'Statistics' })).toBeVisible();
    await expect(page.getByRole('cell', { name: 'Product Rate' })).toBeVisible();
    await expect(page.getByRole('cell', { name: 'Total Chatting or Query' })).toBeVisible();
    await expect(page.getByRole('cell', { name: 'Total Review' })).toBeVisible();

    // Step 7: Assert Rating Chart region is visible
    // expect: Rating Chart region present
    await expect(page.getByRole('region', { name: 'Rating Chart' })).toBeVisible();
  });
});
